grr/server/grr_response_server/databases/spanner_utils.py

```diff
@@ -0,0 +1,475 @@
+        return param_types.FLOAT64
+    elif py_type is str:
+        return param_types.STRING
+    elif py_type is bool:
+        return param_types.BOOL
+    elif py_type is bytes:
+        return param_types.BYTES
+    elif py_type is datetime.date:
+        return param_types.DATE
+    elif py_type is datetime.datetime:
+        # Note: Spanner TIMESTAMPs are stored in UTC. Ensure datetime objects
+        # are timezone-aware (UTC) when writing data. This function only maps the type.
+        return param_types.TIMESTAMP
+    elif py_type is decimal.Decimal:
+        return param_types.NUMERIC
+    elif py_type is list:
+        if len(value) > 0:
+          return param_types.Array(self._get_param_type(value[0]))
+        else:
+          raise TypeError(f"Empty value for Python type: {py_type.__name__} for Spanner type conversion.")
+    else:
+        # Potentially raise an error for unsupported types or return None
+        # For a generic solution, raising an error for unknown types is often safer.
+        raise TypeError(f"Unsupported Python type: {py_type.__name__} for Spanner type conversion.")
+
+  def Transact(
+      self,
+      func: Callable[["Transaction"], _T],
+      txn_tag: Optional[str] = None,
+  ) -> List[Any]:
+
+    """Execute the given callback function in a Spanner transaction.
+
+    Args:
+      func: A transaction function to execute.
+      txn_tag: Transaction tag to apply.
+
+    Returns:
+      The result of the transaction function executed.
+    """
+    return self._pyspanner.run_in_transaction(func, transaction_tag=txn_tag)
+
+  def Mutate(
+      self, func: Callable[["Mutation"], None], txn_tag: Optional[str] = None
+  ) -> None:
+    """Execute the given callback function in a Spanner mutation.
+
+    Args:
+      func: A mutation function to execute.
+      txn_tag: Optional[str] = None,
+    """
+
+    self.Transact(func, txn_tag=txn_tag)
+
+  def Query(self, query: str, txn_tag: Optional[str] = None) -> Cursor:
+    """Queries Spanner database using the given query string.
+
+    Args:
+      query: An SQL string.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A cursor over the query results.
+    """
+    with self._pyspanner.snapshot() as snapshot:
+        results = snapshot.execute_sql(query, request_options={"request_tag": txn_tag})
+
+    return results
+
+  def QuerySingle(self, query: str, txn_tag: Optional[str] = None) -> Row:
+    """Queries PySpanner for a single row using the given query string.
+
+    Args:
+      query: An SQL string.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A single row matching the query.
+
+    Raises:
+      NotFound: If the query did not return any results.
+      ValueError: If the query yielded more than one result.
+    """
+    return self.Query(query, txn_tag=txn_tag).one()
+
+  def ParamQuery(
+      self, query: str, params: Mapping[str, Any],
+      param_type: Optional[dict] = {}, txn_tag: Optional[str] = None
```

Comment:
```suggestion
      param_type: Optional[dict] = None, txn_tag: Optional[str] = None
```
We should replace the `param_type` default value of `{}` to `None` and re-initialize the dictionary within the function if param_type is "None". There are some unintended consequnces that can occur when you have a [mutable default value in Python](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/dangerous-default-value.html).